include/dlaf/matrix/distribution_extensions.h

```diff
@@ -0,0 +1,164 @@
+      distribution.rank_global_tile(distribution.global_tile_index(sub_offset)) ==
+      sub_distribution.source_rank_index();
+  const bool compatible_size =
+      sub_offset.row() + sub_distribution.size().rows() <= distribution.size().rows() &&
+      sub_offset.col() + sub_distribution.size().cols() <= distribution.size().cols();
+  return compatible_grid && compatible_source_rank_index && compatible_size;
+}
+
+/// Returns the tile index in the current distribution corresponding to a tile index @p sub_index in a
+/// sub-distribution (defined by @p sub_offset and @p sub_distribution)
+inline GlobalTileIndex global_tile_index_from_sub_distribution(
+    const Distribution& distribution, const GlobalElementIndex& sub_offset,
+    const Distribution& sub_distribution, const GlobalTileIndex& sub_index) noexcept {
+  DLAF_ASSERT(sub_index.isIn(sub_distribution.nrTiles()), sub_index, sub_distribution.nrTiles());
+  DLAF_ASSERT(is_compatible_sub_distribution(distribution, sub_offset, sub_distribution), "");
+  const GlobalTileIndex tile_offset = distribution.globalTileIndex(sub_offset);
+  return tile_offset + common::sizeFromOrigin(sub_index);
+}
+
+/// Returns the element offset within the tile in the current distribution corresponding to a tile
+/// index @p sub_index in a sub-distribution (defined by @p sub_offset and @p sub_distribution)
+inline TileElementIndex tile_element_offset_from_sub_distribution(
+    const Distribution& distribution, const GlobalElementIndex& sub_offset,
+    const Distribution& sub_distribution, const GlobalTileIndex& sub_index) noexcept {
+  DLAF_ASSERT(sub_index.isIn(sub_distribution.nrTiles()), sub_index, sub_distribution.nrTiles());
+  DLAF_ASSERT(is_compatible_sub_distribution(distribution, sub_offset, sub_distribution), "");
+  return {
+      sub_index.row() == 0 ? distribution.tile_element_from_global_element<Coord::Row>(sub_offset.row())
+                           : 0,
+      sub_index.col() == 0 ? distribution.tile_element_from_global_element<Coord::Col>(sub_offset.col())
+                           : 0,
+  };
+}
+
+/////////////////////////////////////////
+// helpers for tridiagonal eigensolver //
+/////////////////////////////////////////
+
+/// Returns the distance from the global index @p i_gl to the tile adjacent the one containing @p i_gl
+/// along @p rc coordinate.
+template <Coord rc>
+SizeType distance_to_adjacent_tile(const Distribution& dist, SizeType global_element) noexcept {
+  const SizeType global_tile = dist.global_tile_from_global_element<rc>(global_element);
+  const SizeType tile_element = dist.tile_element_from_global_element<rc>(global_element);
+  return dist.size_of_tile<rc>(global_tile) - tile_element;
+}
+
+/// Returns a local linear column-major index of the tile that contains @p ij_local
+inline SizeType local_tile_linear_index(const Distribution& dist, LocalTileIndex ij_local) noexcept {
+  return ij_local.row() + ij_local.col() * dist.local_nr_tiles().rows();
```

Comment:
Using distribution I realised that an assert here would have helped.

```cpp
DLAF_ASSERT(ij_local.isIn(dist.local_nr_tiles()), ij_local, dist.local_nr_tiles());
```

Probably worth having this kind of assert in here more extensively?